mrjones2014/legendary.nvim
# Draft: Version 2 refactor, breaking changes but much better plugin and API

## How to Test

1. Ensure all your keymaps/commands/autocmds/functions still work (after updating your config with any API changes that might be needed)
2. Ensure documentation is accurate for new APIs

## Testing for Regressions

I have tested the following:



Also fixes #60 

# Release Notes

## New Features

- Option to put scratchpad in a floating window
- Command to toggle the scratchpad open and closed (mostly useful if you're using a floating window)
- Ability to define command which accept arguments
- Sorting most recently used items to the top (previously just the single most recent would be at the top)
- Parser to convert Vimscript mapping commands (e.g. `vnoremap <silent> <leader>f :SomeCommand<CR>`) to `legendary.nvim` keymaps
- Based on my testing with [`:StartupTime`](https://github.com/dstein64/vim-startuptime) and [`:PackerProfile`](https://github.com/wbthomason/packer.nvim#profiling), its about twice as fast

## Breaking Changes


- Some config options have moved, see `README.md`
- The `require('legendary.helpers')` module was moved to `require('legendary.toolbox')` and has additional utilities
- The which-key integration functions are no longer on the top-level `require('legendary')`, those are at `require('legendary.integrations.which-key')`, however if `config.which_key.auto_register == true`, they will still be auto-registered
- `kind` field was removed, instead you can use the following functions: 
  - `require('legendary.toolbox').is_keymap()`
  - `require('legendary.toolbox').is_command()`
  - `require('legendary.toolbox').is_func()`
  - `require('legendary.toolbox').is_autocmd()`
  - `require('legendary.toolbox').is_augroup()`
  - `require('legendary.toolbox').is_augroup_or_autocmd()`
- The following functions were renamed:
  - `require('legendary').bind_keymap()` => `require('legendary').keymap()`
  - `require('legendary').bind_keymaps()` => `require('legendary').keymaps()`
  - `require('legendary').bind_command()` => `require('legendary').command()`
  - `require('legendary').bind_commands()` => `require('legendary').commands()`
  - `require('legendary').bind_function()` => `require('legendary').func()`
  - `require('legendary').bind_functions()` => `require('legendary').funcs()`
  - `require('legendary').bind_autocmds()`
    - Split into two functions, `require('legendary').autocmd()` and `require('legendary').autocmds()`
 - Callbacks functions are no longer passed marks as arguments -- this behavior prevented the ability to create commands which accepted arguments themselves, as they would not have been passed to the callback function. Instead, you may use the following utilities:
   - `require('legendary.toolbox').is_visual_mode()`
   - `require('legendary.toolbox').get_marks()`
- Commands are now executed from the picker via `vim.cmd` instead of `vim.api.nvim_feedkeys`; for visual mode mappings with command string implementations, you may need to add marks to your command, for example `: '<,'>SomeCommand<CR>`
